BraveWallet/Extensions/RpcServiceExtensions.swift

```diff
@@ -83,42 +117,109 @@ extension BraveWalletJsonRpcService {
     completion: @escaping (BDouble?) -> Void
   ) {
-    let convert: (String, BraveWallet.ProviderError, String) -> Void = { wei, status, _ in
-      guard status == .success && !wei.isEmpty else {
-        completion(nil)
-        return
-      }
-      let formatter = WeiFormatter(decimalFormatStyle: decimalFormatStyle)
-      if let valueString = formatter.decimalString(
-        for: wei.removingHexPrefix,
-        radix: .hex,
-        decimals: Int(token.decimals)
-      ) {
-        completion(BDouble(valueString))
-      } else {
-        completion(nil)
-      }
-    }
     network(coin) { [self] network in
-      if token.symbol == network.symbol {
-        balance(accountAddress, coin: coin, chainId: network.chainId, completion: convert)
-      } else if token.isErc20 {
-        erc20TokenBalance(
-          token.contractAddress(in: network),
-          address: accountAddress,
-          chainId: network.chainId,
-          completion: convert
+      switch coin {
+      case .eth:
+        ethBalance(
+          for: token,
+          in: accountAddress,
+          network: network,
+          completion: { wei, status, _ in
+            guard status == .success && !wei.isEmpty else {
+              completion(nil)
+              return
+            }
+            let formatter = WeiFormatter(decimalFormatStyle: decimalFormatStyle)
+            if let valueString = formatter.decimalString(
+              for: wei.removingHexPrefix,
+              radix: .hex,
+              decimals: Int(token.decimals)
+            ) {
+              completion(BDouble(valueString))
+            } else {
+              completion(nil)
+            }
+          }
         )
-      } else if token.isErc721 {
-        erc721TokenBalance(
-          token.contractAddress,
-          tokenId: token.tokenId,
-          accountAddress: accountAddress,
-          chainId: network.chainId,
-          completion: convert)
-      } else {
+      case .sol:
+        if token.symbol == network.nativeToken.symbol {
+          solanaBalance(accountAddress, chainId: network.chainId) { lamports, status, errorMessage in
+            guard status == .success else {
+              completion(nil)
+              return
+            }
+            let formatter = WeiFormatter(decimalFormatStyle: decimalFormatStyle)
+            if let valueString = formatter.decimalString(
+              for: "\(lamports)",
+              radix: .decimal,
+              decimals: Int(token.decimals)
+            ) {
+              completion(BDouble(valueString))
+            } else {
+              completion(nil)
+            }
+          }
+        } else {
+          splTokenAccountBalance(
+            accountAddress,
+            tokenMintAddress: token.contractAddress,
+            chainId: network.chainId
+          ) { amount, lamports, uiAmountString, status, errorMessage in
+            guard status == .success else {
```

Comment:
what are `lamports` and `uiAmountString` do here?

Reply:
These are the names from the callback definition, they are different precision of the amount, ex. amount=998999100000000000, lamports=9, uiAmountString=998999100